cibuildwheel/platforms/android.py

```diff
@@ -0,0 +1,625 @@
+    orig_prefix = orig_vars["prefix"]
+    localized_vars_ = {}
+    for key, value in orig_vars.items():
+        # The host's sysconfigdata will include references to build-time paths.
+        # Update these to refer to the current prefix.
+        final = value
+        if isinstance(final, str):
+            final = final.replace(orig_prefix, str(prefix))
+
+        if key == "ANDROID_API_LEVEL":
+            if api_level := build_env.get(key):
+                final = int(api_level)
+
+        # Build systems vary in whether FLAGS variables are read from sysconfig, and if so,
+        # whether they're replaced by environment variables or combined with them. Even
+        # setuptools has changed its behavior here
+        # (https://github.com/pypa/setuptools/issues/4836).
+        #
+        # Ensure consistency by clearing the sysconfig variables and letting the environment
+        # variables take effect alone. This will also work for any non-Python build systems
+        # which the build script may call.
+        elif key in ["CFLAGS", "CXXFLAGS", "LDFLAGS"]:
+            final = ""
+
+        # These variables contain an embedded copy of LDFLAGS.
+        elif key in ["LDSHARED", "LDCXXSHARED"]:
+            final = final.removesuffix(" " + orig_vars["LDFLAGS"])
+
+        localized_vars_[key] = final
+
+    return localized_vars_
+
+
+def setup_android_env(
+    python_dir: Path, venv_dir: Path, build_env: dict[str, str]
+) -> dict[str, str]:
+    site_packages = next(venv_dir.glob("lib/python*/site-packages"))
+    for suffix in ["pth", "py"]:
+        shutil.copy(resources.PATH / f"_cross_venv.{suffix}", site_packages)
+
+    sysconfigdata_path = Path(
+        shutil.copy(
+            next(python_dir.glob("prefix/lib/python*/_sysconfigdata_*.py")),
+            site_packages,
+        )
+    )
+    sysconfigdata = localize_sysconfigdata(python_dir, build_env, sysconfigdata_path)
+
+    android_env = build_env.copy()
+    android_env["CIBW_CROSS_VENV"] = "1"  # Activates the code in _cross_venv.py.
+
+    env_output = call(python_dir / "android.py", "env", env=build_env, capture_stdout=True)
+    for line in env_output.splitlines():
+        key, value = line.removeprefix("export ").split("=", 1)
+        value_split = shlex.split(value)
+        assert len(value_split) == 1, value_split
+        android_env[key] = value_split[0]
+
+    # localized_vars cleared the CFLAGS and CXXFLAGS in the sysconfigdata, but most
+    # packages take their optimization flags from these variables. Pass these flags via
+    # environment variables instead.
+    #
+    # We don't enable debug information, because it significantly increases binary size,
+    # and most Android app developers don't have the NDK installed, so they would have no
+    # way to strip it.
+    opt = " ".join(word for word in sysconfigdata["OPT"].split() if not word.startswith("-g"))
+    for key in ["CFLAGS", "CXXFLAGS"]:
+        android_env[key] += " " + opt
```

Comment:
Do we know that `CFLAGS`/`CXXFLAGS` exists in `android_env`? I'm confused why this line doesn't raise a KeyError.

Oh, I suppose it's always returned by `android.py env`. It might be worth making this more robust though.

Reply:
Yes, `android.py env` always returns all the environment variables necessary for a cross build, including CC, CFLAGS, etc. So I don't think any code changes are required here: if these variables are missing, then something has gone fundamentally wrong, and the KeyError will be as clear an indication as we can give. But I've added a comment explaining what `android.py env` returns, with a link to the source of the variables.